rosidl_parser/rosidl_parser/definition.py

```diff
@@ -0,0 +1,343 @@
+
+    __slots__ = ('name', 'value')
+
+    def __init__(self, name, value):
+        assert isinstance(name, str)
+        self.name = name
+        self.value = value
+
+
+class Annotatable:
+
+    __slots__ = ('annotations', )
+
+    def __init__(self):
+        self.annotations = []
+
+    def get_annotation_value(self, name):
+        values = self.get_annotation_values(name)
+        if not values:
+            raise ValueError("No '{name}' annotation".format_map(locals()))
+        if len(values) > 1:
+            raise ValueError("Multiple '{name}' annotations".format_map(locals()))
+        return values[0]
+
+    def get_annotation_values(self, name):
+        return [a.value for a in self.annotations if a.name == name]
+
+    def has_annotation(self, name):
+        values = self.get_annotation_values(name)
+        return len(values) == 1
+
+    def has_annotations(self, name):
+        annotations = self.get_annotation_values(name)
+        return bool(annotations)
+
+
+class Member(Annotatable):
+
+    __slots__ = ('type', 'name')
+
+    def __init__(self, type_, name):
+        super().__init__()
+        assert isinstance(type_, AbstractType)
+        self.type = type_
+        self.name = name
+
+
+class Structure(Annotatable):
+
+    __slots__ = ('type', 'members')
+
+    def __init__(self, type_):
+        super().__init__()
+        assert isinstance(type_, NamespacedType)
+        self.type = type_
+        self.members = []
+
+
+class Include:
+
+    __slots__ = ('locator', )
+
+    def __init__(self, locator):
+        self.locator = locator
+
+
+class Constant(Annotatable):
+
+    __slots__ = ('name', 'type', 'value')
+
+    def __init__(self, name, type_, value):
+        super().__init__()
+        assert isinstance(type_, AbstractType)
+        self.name = name
+        self.type = type_
+        self.value = value
+
+
+class Message:
+
+    __slots__ = ('structure', 'constants', 'enums')
+
+    def __init__(self, structure):
+        super().__init__()
+        assert isinstance(structure, Structure)
+        self.structure = structure
+        self.constants = OrderedDict()
+        self.enums = []
+
+
+class Service:
+
+    __slots__ = ('structure_type', 'request_message', 'response_message')
+
+    def __init__(self, type_, request, response):
+        super().__init__()
+
+        assert isinstance(type_, NamespacedType)
+        self.structure_type = type_
+
+        assert isinstance(request, Message)
+        assert request.structure.type.namespaces == type_.namespaces
+        assert request.structure.type.name == type_.name + \
+            SERVICE_REQUEST_MESSAGE_SUFFIX
+        self.request_message = request
+
+        assert isinstance(response, Message)
+        assert response.structure.type.namespaces == type_.namespaces
+        assert response.structure.type.name == type_.name + \
+            SERVICE_RESPONSE_MESSAGE_SUFFIX
+        self.response_message = response
+
+
+class IdlIdentifier:
```

Comment:
This doesn't appear to be used. What's it for?

Reply:
Indeed, this is now represented by two separate variables - the package name and the relative path. Removed in f644436528c3469af04e4f325b9b6b65892de6c2.